addons/stock/models/stock_move_line.py

```diff
@@ -456,22 +458,31 @@ def _action_done(self):
             elif qty_done_float_compared < 0:
                 raise UserError(_('No negative quantities allowed'))
             else:
-                ml_to_delete |= ml
+                ml_ids_to_delete.add(ml.id)
+
+        ml_to_delete = self.env['stock.move.line'].browse(ml_ids_to_delete)
         ml_to_delete.unlink()
 
-        (self - ml_to_delete)._check_company()
+        # Batching the creation of lots and associated is to the right ml (order is preverse in the create)
+        lots = self.env['stock.production.lot'].create(lot_vals_to_create)
+        for i, lot in enumerate(lots):
+            associate_line_lot[i].write({'lot_id': lot.id})
+
+        mls_todo = (self - ml_to_delete)
+        mls_todo._check_company()
 
         # Now, we can actually move the quant.
-        done_ml = self.env['stock.move.line']
-        for ml in self - ml_to_delete:
+        ml_ids_to_ignore = OrderedSet()
+        for ml in mls_todo:
```

Comment:
I think I understand what do you want, but unfortunately, it is not possible here:
- About the `_should_bypass_reservation` doesn't depend only of the product but also about the move_id (check mrp_subcontracting). Also it is clearly not a bottleneck (not in my last flamme-graph), Then IMO, it will hit the readability for nothing :/
- "The code is evaluating a bunch of instructions n times ", the problem here is that each instruction depends of the previous one (first sml impact all the rest). Then we can't (trivially) change that (or batching everything but it is quite a complexe problem with quants https://github.com/odoo/odoo/pull/61556).